File: packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderFiles.java
package expo.modules.updates.loader;

import android.content.Context;
import android.util.Log;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.security.NoSuchAlgorithmException;

import expo.modules.updates.UpdatesUtils;
import expo.modules.updates.db.entity.AssetEntity;

/**
 * Utility class for Loader and its subclasses, to allow for easy mocking
 */
public class LoaderFiles {

  private static final String TAG = LoaderFiles.class.getSimpleName();

  public boolean fileExists(File destination) {
    return destination.exists();
  }

  public byte[] copyAssetAndGetHash(AssetEntity asset, File destination, Context context) throws NoSuchAlgorithmException, IOException {
    if (asset.embeddedAssetFilename != null) {
      return copyContextAssetAndGetHash(asset, destination, context);
    } else if (asset.resourcesFilename != null && asset.resourcesFolder != null) {
      return copyResourceAndGetHash(asset, destination, context);
    } else {
      throw new AssertionError("Failed to copy embedded asset " + asset.key + " from APK assets or resources because not enough information was provided.");
    }
  }

  private byte[] copyContextAssetAndGetHash(AssetEntity asset, File destination, Context context) throws NoSuchAlgorithmException, IOException {
    try (
      InputStream inputStream = context.getAssets().open(asset.embeddedAssetFilename)
    ) {
      return UpdatesUtils.sha256AndWriteToFile(inputStream, destination);
    } catch (Exception e) {
      Log.e(TAG, "Failed to copy asset " + asset.embeddedAssetFilename, e);
      throw e;
    }
  }

  private byte[] copyResourceAndGetHash(AssetEntity asset, File destination, Context context) throws NoSuchAlgorithmException, IOException {
    int id = context.getResources().getIdentifier(asset.resourcesFilename, asset.resourcesFolder, context.getPackageName());
    try (
      InputStream inputStream = context.getResources().openRawResource(id)
    ) {
      return UpdatesUtils.sha256AndWriteToFile(inputStream, destination);
    } catch (Exception e) {
      Log.e(TAG, "Failed to copy asset " + asset.embeddedAssetFilename, e);
      throw e;
    }
  }
}

File: packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderFiles.kt
package expo.modules.updates.loader

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log
import expo.modules.updates.UpdatesConfiguration
import expo.modules.updates.UpdatesUtils
import expo.modules.updates.db.entity.AssetEntity
import expo.modules.updates.manifest.EmbeddedManifestUtils
import expo.modules.updates.manifest.Update
import expo.modules.updates.utils.AndroidResourceAssetUtils
import java.io.File
import java.io.IOException
import java.security.NoSuchAlgorithmException

/**
 * Utility class for Loader and its subclasses, to allow for easy mocking
 */
open class LoaderFiles {
  fun fileExists(context: Context, updateDirectory: File?, relativePath: String?): Boolean {
    val filePath = relativePath ?: return false
    if (AndroidResourceAssetUtils.isAndroidAssetOrResourceExisted(context, filePath)) {
      return true
    }
    return File(updateDirectory, filePath).exists()
  }

  fun readEmbeddedUpdate(
    context: Context,
    configuration: UpdatesConfiguration
  ): Update? {
    return EmbeddedManifestUtils.getEmbeddedUpdate(context, configuration)
  }

  @Throws(NoSuchAlgorithmException::class, IOException::class)
  fun copyAssetAndGetHash(asset: AssetEntity, destination: File, context: Context): ByteArray {
    return if (asset.embeddedAssetFilename != null) {
      copyContextAssetAndGetHash(asset, destination, context)
    } else if (asset.resourcesFilename != null && asset.resourcesFolder != null) {
      copyResourceAndGetHash(asset, destination, context)
    } else {
      throw AssertionError("Failed to copy embedded asset " + asset.key + " from APK assets or resources because not enough information was provided.")
    }
  }

  @Throws(NoSuchAlgorithmException::class, IOException::class)
  internal fun copyContextAssetAndGetHash(
    asset: AssetEntity,
    destination: File,
    context: Context
  ): ByteArray {
    try {
      context.assets.open(asset.embeddedAssetFilename!!)
        .use { inputStream -> return UpdatesUtils.verifySHA256AndWriteToFile(inputStream, destination, null) }
    } catch (e: Exception) {
      Log.e(TAG, "Failed to copy asset " + asset.embeddedAssetFilename, e)
      throw e
    }
  }

  @SuppressLint("DiscouragedApi")
  @Throws(NoSuchAlgorithmException::class, IOException::class)
  internal fun copyResourceAndGetHash(
    asset: AssetEntity,
    destination: File,
    context: Context
  ): ByteArray {
    val id = context.resources.getIdentifier(
      asset.resourcesFilename,
      asset.resourcesFolder,
      context.packageName
    )
    try {
      context.resources.openRawResource(id)
        .use { inputStream -> return UpdatesUtils.verifySHA256AndWriteToFile(inputStream, destination, null) }
    } catch (e: Exception) {
      Log.e(TAG, "Failed to copy resource asset ${asset.resourcesFolder}/${asset.embeddedAssetFilename}", e)
      throw e
    }
  }

  companion object {
    private val TAG = LoaderFiles::class.java.simpleName
  }
}
